### PR TITLE
Add benchmarking with JSON Model Compiler and JSON Schema Utils

### PR DESCRIPTION
**Pull request by zx80**
This is a preliminary setup to add [jmc](https://github.com/clairey-zx81/json-model) as an implementation.

- I'm not really sure whether this is enough to run with your benchmarking stuff.
- There are a few diffs, most of which should be resolved by fixing the schemas or the validated instances.
- Overall, I think that the benchmarking runs are too short.

### CODE REVIEW

**Comment by jviotti** [comment]
I'll let @michaelmior comment, as he was the main person setting up this benchmark and its various tricks.

File: implementations/jmc/Dockerfile [hunk 1]
```diff
@@ -0,0 +1,39 @@
...
+# build & utils
+RUN apt install -y --no-install-recommends --fix-missing \
+        vim git ca-certificates libtool m4 time automake make pkgconf gcc g++ texinfo
+
+# libs and all
+RUN apt update
+RUN apt install -y --no-install-recommends --fix-missing \
+        libpcre2-dev libre2-dev libjansson-dev python3 python-is-python3 python3-venv python3-pip
+
+RUN apt clean
```
Comment (michaelmior) [inline implementations/jmc/Dockerfile:16]: Is there any reason all this can't be combined into a single line? See for example the Dockerfile for `ajv-bun`. This would just limit the space taken by intermediate layers.

File: implementations/jmc/Dockerfile [hunk 2]
```diff
@@ -0,0 +1,39 @@
...
+RUN git clone https://github.com/marcomaggi/cre2.git
+RUN cd cre2 ; \
+    sh autogen.sh ; \
+    ./configure --enable-maintainer-mode ; \
+    make ; \
+    make install
+
+# setup python environment
+RUN python -m venv /venv
+RUN /venv/bin/pip install -U pip
```
Comment (michaelmior) [inline implementations/jmc/Dockerfile:28]: Unless this causes specific problems, I would suggest not using a virtualenv here since Docker already isolates everything.

File: implementations/jmc/Makefile
Comment (michaelmior) [inline implementations/jmc/Makefile:1]: I don't really understand the purpose of a separate Makefile here. It appears to duplicate quite a bit of the logic from the main Makefile.

File: implementations/jmc/README.md [hunk 3]
```diff
@@ -0,0 +1,110 @@
...
+- [x] stylecop pass=983 fail=0 error=0
+- [x] tmuxinator pass=382 fail=0 error=0
+- [x] ui5 pass=942 fail=0 error=0
+- [x] ui5-manifest **pcre2** pass=611 fail=0 error=0
+  - one regular expressions extension incompatible with re2
+- [x] unreal-engine-uproject pass=859 fail=0 error=0
+- [x] vercel pass=710 fail=0 error=0
+- [x] yamllint pass=966 fail=18 error=0
+  - the 18 differences are raw strings, model conversion rightfully infer object requirement
+    thus reject strings like "hello world" as valid yamllint values
```
Comment (michaelmior) [inline implementations/jmc/README.md:85]: I would suggest not putting such detailed status information here as it can quickly get out of date.

File: implementations/jmc/README.md [hunk 4]
```diff
@@ -0,0 +1,110 @@
...
+
+```sh
+make BENCH=lerna bench  # run for one benchmark directory
+make benches            # run all benchmarks
+make benches.txt        # run for all benchmark directories, with results in this file
+make clean              # local cleanup
+make clean.docker       # more cleanup
+make version            # show jmc versions
+make start/stop         # docker container operations
+```
```
Comment (michaelmior) [inline implementations/jmc/README.md:97]: See my comments above about the Makefile.

File: implementations/jmc/generate-and-run.sh [hunk 5]
```diff
@@ -0,0 +1,98 @@
...
+H run time: $(( $run_time / 1000 )) µs
+
+# run again with internally measured validation time
+./model.out --jsonschema-benchmark -T $LOOP "$INSTANCES" > time.txt
+let validation_time=$(cat time.txt)
+
+H validation time: $(( $validation_time / 1000 )) µs
+
+# timing & exit
+echo $validation_time,$validation_time,$compile_time
```
Comment (michaelmior) [inline implementations/jmc/generate-and-run.sh:97]: Like all other implementations, there should be two separate validation time measurements. Once for a single pass and another after running the validation for several times.

**Review by michaelmior — CHANGES_REQUESTED** [review]
Thanks for the additional implementation! There are a few things that I think should be cleaned up and the GitHub Action configuration also needs to be updated for this to run on CI.

File: implementations/jmc/Makefile
Comment (zx80) [reply implementations/jmc/Makefile:1]: This makefile is more convenient for me to run a lot of local tests, not just benchmarking. I may need it in the future, so I'd like to keep it there.

File: implementations/jmc/README.md (hunk 3 above)
Comment (zx80) [reply implementations/jmc/README.md]: Sure, this was for recording the work-in-progress. I've removed details and just kept the current issues.

**Comment by zx80** [comment]
The directory is currently named `jmc` and it tests the C backend. I will probably submit other entries for JS and Python backends. Should the directory be named `jmc-c` to reflect the underlying library?

**Review by michaelmior — CHANGES_REQUESTED** [review]
Thanks for the changes. I left a few additional comments.

File: implementations/jmc/Dockerfile [hunk 6]
```diff
@@ -23,12 +19,12 @@ RUN cd cre2 ; \
...
     make install
 
-# setup python environment
+# setup python environment, native pip complains about direct installations otherwise
 RUN python -m venv /venv
 RUN /venv/bin/pip install -U pip
 
 # direct install json-model and json-schema-utils from github
-RUN echo 33
+RUN echo jmc 0
```
[the remark this answers is not in the document]
Comment (zx80) [reply implementations/jmc/Dockerfile]: Hmmm… it helps to work easily around docker _clever_ caching while testing, but in the long end it is indeed probably not necessary.

File: implementations/jmc/Dockerfile [hunk 7]
```diff
@@ -1,18 +1,14 @@
 FROM ubuntu
```
[the remark this answers is not in the document]
Comment (zx80) [reply implementations/jmc/Dockerfile:1]: I try to avoid hardcoding versions in general, but I can put one anyway.
My own environment is ubuntu and I know my way around it, so it helps _me_ to have ubuntu there as well.

**Comment by zx80** [comment]
I think I did what you asked. I've added entries for js and python backends.

File: implementations/jmc/Dockerfile (hunk 7 above)
[the remark this answers is not in the document]
Comment (michaelmior) [reply implementations/jmc/Dockerfile]: @zx80 The reason we want versions hardcoded is to make things as reproducible as possible. It also makes maintenance easier since things are less likely to break in the future with some new Ubuntu release. We can manually decide when an upgrade is desired/necessary. Until then, things should continue functioning more or less indefinitely.

File: implementations/jmc/generate-and-run.sh [hunk 8]
```diff
@@ -0,0 +1,134 @@
...
+$bench --jsonl "$INSTANCES" > model.txt
+status=$?
+let run_end=$(date +%s.%N | tr -d .)
+let run_time=$(( $run_end - $run_start))
+
+pass=$(grep PASS model.txt | wc -l)
+fail=$(grep FAIL model.txt | wc -l)
+err=$(grep ERROR model.txt | wc -l)
+
+H results: pass=$pass fail=$fail error=$err
```
Comment (michaelmior) [inline implementations/jmc/generate-and-run.sh:123]: Note that if there are any failures, the process should exit with a non-zero status code.

File: implementations/jmc/generate-and-run.sh [hunk 9]
```diff
@@ -89,10 +90,10 @@ H run time: $(( $run_time / 1000 )) µs
...
 # run again with internally measured validation time
 ./model.out --jsonschema-benchmark -T $LOOP "$INSTANCES" > time.txt
-let validation_time=$(cat time.txt)
 
+let validation_time=$(cat time.txt | cut -d, -f2)
 H validation time: $(( $validation_time / 1000 )) µs
 
 # timing & exit
-echo $validation_time,$validation_time,$compile_time
+echo $(cat time.txt| tr -d '\012'),$compile_time
```
[the remark this answers is not in the document]
Comment (michaelmior) [reply implementations/jmc/generate-and-run.sh:133]: It's hard for me to see that runtime measurements are consistent with how other implementations are measured. There are two measurements. The first validates all documents a single time. Then, documents are validated several more times without measurement. Finally, each document is validated one more time and this total time is measured.

**Review by michaelmior — CHANGES_REQUESTED** [review]
Thanks for addressing those changes! My final concern is making sure that how things are measured is consistent with existing implementations.

File: implementations/jmc/generate-and-run.sh (hunk 8 above)
Comment (zx80) [reply implementations/jmc/generate-and-run.sh:133]: It should, because the commands status recovered on line 115 is non zero if not eveything passed, and is used to exit on line 134.

Anyway, I added  a consistency check with the output when the first command claims success, and another status check on the actual benchmarking run.

**Comment by zx80** [comment]
Hello,

> It's hard for me to see that runtime measurements are consistent with how other implementations are measured. There are two measurements. The first validates all documents a single time. Then, documents are validated several more times without measurement. Finally, each document is validated one more time and this total time is measured.

That is _more-or-less_ what is done by each implementation, as can be seen from source codes:

- [C backend](https://github.com/clairey-zx81/json-model/blob/main/json_model/runtime/main.c#L162)
- [Python backend](https://github.com/clairey-zx81/json-model/blob/main/json_model/runtime/support.py#L343)
- [JS backend](https://github.com/clairey-zx81/json-model/blob/main/js_runtime/main.js#L65)

There is no intermediate warming runs on C code because I do not expect anything there, but I can add it if this is an absolute requirement. I've just added one for Python, but I do not expect it to do anything, the internal JIT if any should have been triggered in the cold loop. There was already one in JS because your code had one and I replicated it.

Further caveats:

- I do **not** trust single run measures, especially from languages which have a garbage collector, because garbage collection may or may not have occured, which makes the result intrinsically unstrusworthy. The second measure in my implementations is an average over _100_ runs, with the _naive_ hope that some gc time should be included. You may notice that standard deviation is quite high for JS, which is a hint that this strategy may be working, and that single run measures are not to be trusted.
- Even with a loop, I'm wary of the reliability and fairness of these measures: the gc may run in a separate thread,which is great for latency (which is what is measured) but not for throuput (what seems more interesting for the use case).
- It is unclear to me whether the various measures from different languages collect CPU time or WALLCLOCK times, the benchmarks may compare different kinds of time. ISTM that the benchmark should be interested in CPU time.
- Given the _crazy things_ that docker does, it is unclear to me whether I should trust performance figures in such an environment. It is probably okay for cpu-only stuff.
- When benchmarking, the _power mode_ of my laptop and the fact that it has heterogeneous cores (2 performance cores, 8 efficient cores) and a frequency varying from 0.8 GHz to 4.7 GHz makes performance results quite unreliable.
- Because there is no failure tests to check whether something is actually validated, `return true` would achieve both perfect performance and reliability: any issue in the tool chains may go undetected.

File: implementations/jmc/generate-and-run.sh (hunk 9 above)
[the remark this answers is not in the document]
Comment (zx80) [reply implementations/jmc/generate-and-run.sh:149]: Done like that in updated runtimes.

**Comment by zx80** [comment]
@michaelmior I'm unclear wether any input is still needed from my part or if the PR is in you hands only. Let me know!

**Comment by zx80** [comment]
FIY I justed force-pushed a rebase.

**Comment by michaelmior** [comment]
@jviotti It looks like Blaze is failing now. Can you take a look?

**Comment by zx80** [comment]
@michaelmior I just did a rebase and added entries for the jmc implementations in the plot/table generation scripts.

Please tell me if there is anything I can do to help moving this PR forward.

**Comment by zx80** [comment]
Indeed. Yes. There is an issue in the csv file contents, having a direct look would have helped. I'll investigate as soon as possible.

**Comment by zx80** [comment]
After some testing, my (blind) guess is that there is a timeout on some perl runs, because the implementation is very slow, it is more of a poc. My (blind) fix is to reduce the loop, which may or may not be enough as a fix.

ISTM that the csv parsing should be more resilient and catch parse errors (in my local test one csv field in one file was empty) and either ignore the entry or maybe put some bottom value?

**Comment by zx80** [comment]
The CI failed in the _aggregate_ task, inside _Output Summary_:

> Error: An error occurred trying to start process '/usr/bin/bash' with working directory '/home/runner/work/jsonschema-benchmark/jsonschema-benchmark'. Argument list too long

 The output is too long? The summary is too large. The issue does not seem to be with the PR itself.

**Comment by zx80** [comment]
```yaml
run: python -c 'import os; print(os.environ["MARKDOWN_TABLE"])' >> $GITHUB_STEP_SUMMARY
```

I can do that, but I'd be surprised if it works:
- its unclear to me whether the issue is triggered before or after executing the command, it may be while processing the environment _before_ command execution.
- the issue does not seem to be about the number of arguments to `echo`, but the size and possibly structure of the variable contents when processed downstream by some unknown github pre/post-processing phase? 
- the idea of storing a significant amount of data in environment variables and passing them as parameters does not seem to scale well: a quick search seems to suggest that  exec limit size is about 2 MB, and that each argument may be limited to 132 KB, for instance. There may be other limits in place.

Note that the _markdown table created_ is already shown with the previous item, so I'm not sure of the purpose of this echo anyway. As this is non essential, I simply commented out the entry to see what happens.

**Comment by zx80** [comment]
The new run failed while uploading plots later on:
 ```
Error: Error: Request failed with status code 429
```

HTTP _429_ is _too many requests_, which seems to be a load issue… I'm not sure what I can do about that from the PR.

**Comment by zx80** [comment]
> @zx80 Looks like it failed again. I'll try again, but it looks like everything with your changes should be fine. Now that the changes are finalized, would you mind rebasing everything into a single commit?

Done.

Note that the Perl implementation is _very_ slow so may not be worth running by default.

**Comment by zx80** [comment]
@michaelmior @jviotti congratulation for your paper acceptance!

About this PR, can I do anything to help processing it?

**Comment by zx80** [comment]
The run still fails when uploading the plots to **ingur** on a _429_ apparent load issue.

I've pushed an update with the upload split in 2.

**Comment by michaelmior** [comment]
It seems like the upload worked, but you'll also need two instances of the "Add plots to summary" step.

**Comment by michaelmior** [comment]
@zx80 It will need to be two separate steps as the code parses the input as JSON. So with a single step it would be `IMG_URLS='[…,…][…,…]' which is not valid JSON.

**Comment by zx80** [comment]
Indeed, the scripts needed updating. I tried to add a `-l` option to handle a list of files instead of a directory/file with an optional name.

**Comment by zx80** [comment]
This does not seem strictly related to the 4 implementations added by the patch.

Should I remove the unsuccessful changes on the CI configuration and submit a patch which only includes the implementations, which may be committed independently of CI improvements?

**Comment by michaelmior** [comment]
@zx80 Indeed. Addressing this separately would be my preference.

**Comment by michaelmior** [comment]
@zx80 Can you rebase on top of the main branch? The upload should be fixed.

**Comment by zx80** [comment]
Hmmm… the S3 upload failed with a strange output, it seems that the environment did not contain what was expected?
